src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -125,12 +182,12 @@ public async Task ProcessRequestsAsync<TContext>(IHttpApplication<TContext> appl
                     return;
                 }
 
-                if (!_stopping)
+                if (_connectionState != ConnectionState.Closed)
                 {
                     await _frameWriter.WriteSettingsAsync(_serverSettings);
                 }
 
-                while (!_stopping)
+                while (_connectionState == ConnectionState.Open || (_connectionState == ConnectionState.Closing && !_streams.IsEmpty))
```

Comment:
Since you are predicating this condition both on the connection state and the emptiness of the stream, you need to an acquire a lock in order to prevent a race where the stream count goes from zero to non-zero while the connection is in the closing state.

The two ways you could do this that I see is by either putting `_connectionState == ConnectionState.Closing && !_streams.IsEmpty` inside a property and acquiring the lock. That seems pretty heavy handed for something that's checked every frame though.

It's probably better to just move `_streams.TryRemove(streamId, out _);` from OnStreamCompleted inside the lock. I think this would make it impossible for the stream count goes from zero to non-zero while the connection is in the closing state.f

Reply:
I just thought about this a bit more. Why not make this just `while (_connectionState != ConnectionState.Closed)` since we call CancelPendingRead in OnStreamCompleted anyway? I still like calling `_streams.TryRemove` inside the lock, but it shouldn't be relevant for this condition if we simplify it.